packages/convex-helpers/server/zod4.ts

```diff
@@ -0,0 +1,642 @@
+}
+
+function makeUnion(members: GenericValidator[]) {
+  const arr = members.filter((m): m is GenericValidator => !!m);
+  if (arr.length === 0) return v.any();
+  if (arr.length === 1) return arr[0];
+  const [a, b, ...rest] = arr as [GenericValidator, GenericValidator, ...GenericValidator[]];
+  return v.union(a, b, ...rest);
+}
+
+// Narrow helpers for Convex validators
+function asValidator(x: unknown): any {
+  return x as unknown as any;
+}
+
+// Zod v4 -> Convex (input)
+function zodToConvexInternal(
+  zodValidator: z.ZodTypeAny,
+  visited: Set<z.ZodTypeAny> = new Set(),
+): GenericValidator {
+  if (!zodValidator) return v.any();
+  if (visited.has(zodValidator)) return v.any();
+  visited.add(zodValidator);
+
+  if (isZid(zodValidator)) {
+    const meta = registryHelpers.getMetadata(zodValidator);
+    return v.id((meta?.tableName as string) ?? "unknown");
+  }
+
+  if (zodValidator instanceof z.ZodDefault) {
+    const inner = zodValidator.unwrap() as unknown as z.ZodTypeAny;
+    const innerV = zodToConvexInternal(inner, visited);
+    return v.optional(asValidator(innerV));
+  }
+  if (zodValidator instanceof z.ZodOptional) {
+    const inner = zodValidator.unwrap() as unknown as z.ZodTypeAny;
+    const innerV = zodToConvexInternal(inner, visited);
+    return v.optional(asValidator(innerV));
+  }
+  if (zodValidator instanceof z.ZodNullable) {
+    const inner = zodValidator.unwrap() as unknown as z.ZodTypeAny;
+    const innerV = zodToConvexInternal(inner, visited);
+    return v.union(asValidator(innerV), v.null());
+  }
+
+  if (zodValidator instanceof z.ZodString) return v.string();
+  if (zodValidator instanceof z.ZodNumber) return v.float64();
+  if (zodValidator instanceof z.ZodBigInt) return v.int64();
+  if (zodValidator instanceof z.ZodBoolean) return v.boolean();
+  if (zodValidator instanceof z.ZodNull) return v.null();
+  if (zodValidator instanceof z.ZodAny) return v.any();
+  if (zodValidator instanceof z.ZodUnknown) return v.any();
+  if (zodValidator instanceof z.ZodDate) return v.float64();
+
+  if (zodValidator instanceof z.ZodArray) {
+    const el = zodToConvexInternal(zodValidator.element as unknown as z.ZodTypeAny, visited);
+    return v.array(asValidator(el));
+  }
+
+  if (zodValidator instanceof z.ZodObject) {
+    const shape = zodValidator.shape as unknown as Record<string, z.ZodTypeAny>;
+    const out: Record<string, GenericValidator> = {};
+    for (const [k, val] of Object.entries(shape)) {
+      out[k] = zodToConvexInternal(val as z.ZodTypeAny, visited);
+    }
+    return v.object(out);
+  }
+
+  if (zodValidator instanceof z.ZodUnion) {
+    const options = (zodValidator as any).options as z.ZodTypeAny[];
+    if (Array.isArray(options) && options.length > 0) {
+      const members = options.map((opt) => zodToConvexInternal(opt, visited));
+      return (makeUnion(members as [GenericValidator, GenericValidator, ...GenericValidator[]]) as unknown) as GenericValidator;
```

Comment:
Why are the casts necessary here?
```suggestion
      return makeUnion(members);
```